### PR TITLE
:earth_africa:  Wrapping asio lib to create a custom network library (EPI-40)

### PR DESCRIPTION
**Pull request by 6im0n**
This pull request introduces significant changes to the R-Type-Reborn project, focusing on adding network capabilities using the Asio library, restructuring the project, and providing documentation. Key changes include adding client-server network functionality, updating the build configuration, and including a detailed installation guide for `vcpkg`.

### Network Functionality:

* `src/network/client/Client.cpp` and `src/network/client/Client.hpp`: Implemented a `Client` class that handles connecting, sending, receiving, and disconnecting from a server using both TCP and UDP protocols. [[1]](diffhunk://#diff-ca0eef843161a1a7341a6970a29de77ba6c52acb60182df7931461cc082e51cdR1-R83) [[2]](diffhunk://#diff-82184981fbfd14bdc53c351dc1bc4a3f0db0da15e21b44259a6bd1299a2235efR1-R106)
* `src/network/server/Server.cpp` and `src/network/server/Server.hpp`: Implemented a `Server` class that handles starting, stopping, sending, and receiving data, as well as managing client connections. [[1]](diffhunk://#diff-d085b2cdabf72993c27ef0e4e28235e9f0a4d95eba20d3ec2ebccdbf9d04cc55R1-R108) [[2]](diffhunk://#diff-ce82f6a0019e8ee9c5f566fc88de94cf71cc814f8a7e30467e50990c0f1c7117R1-R134)

### Build Configuration:

* [`CMakeLists.txt`](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR2-R24): Updated the build configuration to include the Asio library and added subdirectories for network components.
* [`src/network/CMakeLists.txt`](diffhunk://#diff-eac9acce65b7952c47b3eb85578f05e7940aca43de1c1c2d202a20e27adb9406R1-R16): Added CMake configuration for building the network library, including linking with Asio.

### Documentation: (outdated)

* [`documentation/vcpkg_installation_guide.md`](diffhunk://#diff-df2755d296edc179d085886168d5e37aa7bc2887b151c4a44553df3298f543ebR1-R106): Added a comprehensive guide for installing `vcpkg` on Windows, Linux, and macOS.

### Code Cleanup:

* [`src/main.cpp`](diffhunk://#diff-34d21af3c614ea3cee120df276c9c4ae95053830d7f1d3deaf009a4625409ad2L1-L14): Removed the old main file and its associated code.
* [`src/CMakeLists.txt`](diffhunk://#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4L1-L5): Removed outdated source references.

### Test Files:

* [`tests/client/main.cpp`](diffhunk://#diff-a10939fe2594bfc711d6cc04c20f2b59ee86c350c13973fced3c50ebc58841d4R1-R50): Added a test client that demonstrates connecting to a server and sending a message.
* [`tests/server/main.cpp`](diffhunk://#diff-35a05a3da403f39b990969e0102370c5ca3447343c869d5a689305b77fe53951R1-R38): Added a test server that demonstrates starting the server and handling incoming connections.

### CODE REVIEW

**Review by landryarki — APPROVED** [review]
:100:

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,3 +14,7 @@ find_package(asio REQUIRED)
 
 add_subdirectory(client)
 add_subdirectory(server)
+add_subdirectory(network)
+
+include_directories(network)
+#add_subdirectory(tests)
```
Comment (Marius-P1) [inline CMakeLists.txt]:
```suggestion
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -13,3 +14,7 @@ find_package(asio REQUIRED)
 
 add_subdirectory(client)
 add_subdirectory(server)
+add_subdirectory(network)
+
+include_directories(network)
```
Comment (Marius-P1) [inline CMakeLists.txt]: I think that should be before the add_sub client and server

File: network/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (Marius-P1) [inline network/CMakeLists.txt]: You don't need to write it again, it's already in the root CMakeLists
```suggestion
```

File: network/client/Client.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    _id = -1;
+    _send_timer = std::make_shared<asio::steady_timer>(_io_context, std::chrono::milliseconds(15));
+    _recv_buffer.resize(2048);
+}
+
+
+
+void Network::Client::connect()
+{
+
```
Comment (Marius-P1) [inline network/client/Client.cpp]:
```suggestion
```

File: network/client/Client.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    // TCP connection init
+    asio::ip::tcp::resolver resolver(_io_context);
+    asio::ip::tcp::resolver::results_type const endpoints = resolver.resolve(_host, std::to_string(_TCP_PORT));
+    asio::error_code error;
+    _tcp_endpoint = asio::connect(_tcp_socket, endpoints, error);
+    if (error) {
+        std::cerr << "Error: " << error.message() << std::endl;
+        return;
```
Comment (Marius-P1) [inline network/client/Client.cpp]: All your error should use throw (except if it's not a critical error)
Create a NetworkException that inherit from IError which is present in includes/stellarforge/common

File: network/client/Client.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+    asio::read(_tcp_socket, asio::buffer(&_id, sizeof(_id)), error);
+    if (error) {
+        std::cerr << "Error: " << error.message() << std::endl;
+        return;
```
Comment (Marius-P1) [inline network/client/Client.cpp]: All your error should use throw (except if it's not a critical error)

File: network/server/InternalMessage/ClientConnected.hpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+     * @brief Gets the connection message
+     * @return The connection message
+     * @version 0.1.0
+     * @since 0.1.0
+     */
+    [[nodiscard]] std::string getMessage() const override;
+
+private:
+    std::string _message; ///< The connection message
+    std::uint8_t _id; ///< The client ID
```
Comment (Marius-P1) [inline network/server/InternalMessage/ClientConnected.hpp:59]: Why not use int

File: network/server/Server.cpp [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    _io_context.run(); // run the io_context event loop for asynchronous operations (non-blocking)
+}
+
+void Network::Server::connect_new_client() {
+    auto socket = std::make_shared<asio::ip::tcp::socket>(_io_context);
+
+    // Accept new client connection in TCP
+    _acceptor.async_accept(*socket, [this, socket](const asio::error_code& error) {
+        if (!error) {
+            int8_t id = 0;
```
Comment (Marius-P1) [inline network/server/Server.cpp]: Use int for id
```suggestion
            int id = 0;
```

File: server/CMakeLists.txt [hunk 9]
```diff
@@ -5,6 +5,7 @@ add_executable(r-type_server)
 
 target_link_libraries(r-type_server PUBLIC asio::asio)
 
+
```
Comment (Marius-P1) [inline server/CMakeLists.txt:8]:
```suggestion
```

File: server/CMakeLists.txt [hunk 10]
```diff
@@ -21,3 +22,5 @@ target_sources(r-type_server
         ${CMAKE_CURRENT_SOURCE_DIR}/GameLogic.cpp
         ${CMAKE_CURRENT_SOURCE_DIR}/utils/BinaryVector.cpp
 )
+
+target_link_libraries(r-type_server PRIVATE rtype::network)
```
Comment (Marius-P1) [inline server/CMakeLists.txt]: This should be under the other target_link_libraries and not under the source files

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
Some little changes

**Review by Marius-P1 — APPROVED** [review]
:ok_hand:

**Pull request merged**